include/onnxruntime/core/framework/data_types.h

```diff
@@ -45,13 +46,76 @@ union MLFloat16 {
   MLFloat16() : val(0) {}
 };
 
-inline bool operator==(const MLFloat16& left, const MLFloat16& right)
-{
+inline bool operator==(const MLFloat16& left, const MLFloat16& right) {
   return left.val == right.val;
 }
 
-inline bool operator!=(const MLFloat16& left, const MLFloat16& right)
-{
+inline bool operator!=(const MLFloat16& left, const MLFloat16& right) {
+  return left.val != right.val;
+}
+
+struct ort_endian {
+  union q {
+    uint16_t v_;
+    uint8_t b_[2];
+    constexpr q(uint16_t v) noexcept : v_(v) {}
+  };
+  static constexpr bool is_little() {
+    return q(0x200).b_[0] == 0x0;
+  }
+  static constexpr bool is_big() {
+    return q(0x200).b_[0] == 0x2;
+  }
+};
+
+//BFloat16
+struct BFloat16 {
+  uint16_t val;
+  explicit BFloat16() : val(0) {}
+  explicit BFloat16(uint16_t v) : val(v) {}
+  explicit BFloat16(float v) {
+    uint16_t* dst = reinterpret_cast<uint16_t*>(&v);
+    if (ort_endian::is_little()) {
```

Comment:
why don't we require these checks for MLFloat16?

Reply:
Because the conversion to MLFloat16 happens by other means. BFLoat16 just drops a byte and we need to know the correct byte to drop or assign to. MlFloat16 is a separate format that requires exponent/mantissa manipulation and is done by a special rather complex Eigen library call.

---
In reply to: [243457073](https://github.com/Microsoft/onnxruntime/pull/235#discussion_r243457073) [](ancestors = 243457073)